// @theia/monaco-editor-core/esm/vs/editor/browser/widget/diffEditor/decorations.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.arrowRevertChange = exports.diffDeleteDecorationEmpty = exports.diffWholeLineDeleteDecoration = exports.diffDeleteDecoration = exports.diffAddDecorationEmpty = exports.diffWholeLineAddDecoration = exports.diffAddDecoration = exports.diffLineDeleteDecorationBackground = exports.diffLineAddDecorationBackground = exports.diffLineDeleteDecorationBackgroundWithIndicator = exports.diffLineAddDecorationBackgroundWithIndicator = exports.diffRemoveIcon = exports.diffInsertIcon = void 0;
const codicons_1 = require("../../../../base/common/codicons");
const htmlContent_1 = require("../../../../base/common/htmlContent");
const themables_1 = require("../../../../base/common/themables");
const textModel_1 = require("../../../common/model/textModel");
const nls_1 = require("../../../../nls");
const iconRegistry_1 = require("../../../../platform/theme/common/iconRegistry");
exports.diffInsertIcon = (0, iconRegistry_1.registerIcon)('diff-insert', codicons_1.Codicon.add, (0, nls_1.localize)('diffInsertIcon', 'Line decoration for inserts in the diff editor.'));
exports.diffRemoveIcon = (0, iconRegistry_1.registerIcon)('diff-remove', codicons_1.Codicon.remove, (0, nls_1.localize)('diffRemoveIcon', 'Line decoration for removals in the diff editor.'));
exports.diffLineAddDecorationBackgroundWithIndicator = textModel_1.ModelDecorationOptions.register({
    className: 'line-insert',
    description: 'line-insert',
    isWholeLine: true,
    linesDecorationsClassName: 'insert-sign ' + themables_1.ThemeIcon.asClassName(exports.diffInsertIcon),
    marginClassName: 'gutter-insert',
});
exports.diffLineDeleteDecorationBackgroundWithIndicator = textModel_1.ModelDecorationOptions.register({
    className: 'line-delete',
    description: 'line-delete',
    isWholeLine: true,
    linesDecorationsClassName: 'delete-sign ' + themables_1.ThemeIcon.asClassName(exports.diffRemoveIcon),
    marginClassName: 'gutter-delete',
});
exports.diffLineAddDecorationBackground = textModel_1.ModelDecorationOptions.register({
    className: 'line-insert',
    description: 'line-insert',
    isWholeLine: true,
    marginClassName: 'gutter-insert',
});
exports.diffLineDeleteDecorationBackground = textModel_1.ModelDecorationOptions.register({
    className: 'line-delete',
    description: 'line-delete',
    isWholeLine: true,
    marginClassName: 'gutter-delete',
});
exports.diffAddDecoration = textModel_1.ModelDecorationOptions.register({
    className: 'char-insert',
    description: 'char-insert',
    shouldFillLineOnLineBreak: true,
});
exports.diffWholeLineAddDecoration = textModel_1.ModelDecorationOptions.register({
    className: 'char-insert',
    description: 'char-insert',
    isWholeLine: true,
});
exports.diffAddDecorationEmpty = textModel_1.ModelDecorationOptions.register({
    className: 'char-insert diff-range-empty',
    description: 'char-insert diff-range-empty',
});
exports.diffDeleteDecoration = textModel_1.ModelDecorationOptions.register({
    className: 'char-delete',
    description: 'char-delete',
    shouldFillLineOnLineBreak: true,
});
exports.diffWholeLineDeleteDecoration = textModel_1.ModelDecorationOptions.register({
    className: 'char-delete',
    description: 'char-delete',
    isWholeLine: true,
});
exports.diffDeleteDecorationEmpty = textModel_1.ModelDecorationOptions.register({
    className: 'char-delete diff-range-empty',
    description: 'char-delete diff-range-empty',
});
exports.arrowRevertChange = textModel_1.ModelDecorationOptions.register({
    description: 'diff-editor-arrow-revert-change',
    glyphMarginHoverMessage: new htmlContent_1.MarkdownString(undefined, { isTrusted: true, supportThemeIcons: true })
        .appendMarkdown((0, nls_1.localize)('revertChangeHoverMessage', 'Click to revert change')),
    glyphMarginClassName: 'arrow-revert-change ' + themables_1.ThemeIcon.asClassName(codicons_1.Codicon.arrowRight),
    zIndex: 10001,
});
