// @theia/monaco-editor-core/esm/vs/editor/contrib/toggleTabFocusMode/browser/toggleTabFocusMode.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToggleTabFocusModeAction = void 0;
const aria_1 = require("../../../../base/browser/ui/aria/aria");
const tabFocus_1 = require("../../../browser/config/tabFocus");
const nls = require("../../../../nls");
const actions_1 = require("../../../../platform/actions/common/actions");
class ToggleTabFocusModeAction extends actions_1.Action2 {
    constructor() {
        super({
            id: ToggleTabFocusModeAction.ID,
            title: { value: nls.localize({ key: 'toggle.tabMovesFocus', comment: ['Turn on/off use of tab key for moving focus around VS Code'] }, 'Toggle Tab Key Moves Focus'), original: 'Toggle Tab Key Moves Focus' },
            precondition: undefined,
            keybinding: {
                primary: 2048 /* KeyMod.CtrlCmd */ | 43 /* KeyCode.KeyM */,
                mac: { primary: 256 /* KeyMod.WinCtrl */ | 1024 /* KeyMod.Shift */ | 43 /* KeyCode.KeyM */ },
                weight: 100 /* KeybindingWeight.EditorContrib */
            },
            f1: true
        });
    }
    run() {
        const oldValue = tabFocus_1.TabFocus.getTabFocusMode();
        const newValue = !oldValue;
        tabFocus_1.TabFocus.setTabFocusMode(newValue);
        if (newValue) {
            (0, aria_1.alert)(nls.localize('toggle.tabMovesFocus.on', "Pressing Tab will now move focus to the next focusable element"));
        }
        else {
            (0, aria_1.alert)(nls.localize('toggle.tabMovesFocus.off', "Pressing Tab will now insert the tab character"));
        }
    }
}
exports.ToggleTabFocusModeAction = ToggleTabFocusModeAction;
ToggleTabFocusModeAction.ID = 'editor.action.toggleTabFocusMode';
(0, actions_1.registerAction2)(ToggleTabFocusModeAction);
