// @theia/monaco-editor-core/esm/vs/editor/contrib/inlineCompletions/browser/commands.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToggleAlwaysShowInlineSuggestionToolbar = exports.HideInlineCompletion = exports.AcceptInlineCompletion = exports.AcceptNextLineOfInlineCompletion = exports.AcceptNextWordOfInlineCompletion = exports.TriggerInlineSuggestionAction = exports.ShowPreviousInlineSuggestionAction = exports.ShowNextInlineSuggestionAction = void 0;
const observable_1 = require("../../../../base/common/observable");
const editorExtensions_1 = require("../../../browser/editorExtensions");
const editorContextKeys_1 = require("../../../common/editorContextKeys");
const commandIds_1 = require("./commandIds");
const inlineCompletionContextKeys_1 = require("./inlineCompletionContextKeys");
const inlineCompletionsController_1 = require("./inlineCompletionsController");
const suggest_1 = require("../../suggest/browser/suggest");
const nls = require("../../../../nls");
const actions_1 = require("../../../../platform/actions/common/actions");
const configuration_1 = require("../../../../platform/configuration/common/configuration");
const contextkey_1 = require("../../../../platform/contextkey/common/contextkey");
class ShowNextInlineSuggestionAction extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: ShowNextInlineSuggestionAction.ID,
            label: nls.localize('action.inlineSuggest.showNext', "Show Next Inline Suggestion"),
            alias: 'Show Next Inline Suggestion',
            precondition: contextkey_1.ContextKeyExpr.and(editorContextKeys_1.EditorContextKeys.writable, inlineCompletionContextKeys_1.InlineCompletionContextKeys.inlineSuggestionVisible),
            kbOpts: {
                weight: 100,
                primary: 512 /* KeyMod.Alt */ | 94 /* KeyCode.BracketRight */,
            },
        });
    }
    run(accessor, editor) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const controller = inlineCompletionsController_1.InlineCompletionsController.get(editor);
            (_a = controller === null || controller === void 0 ? void 0 : controller.model.get()) === null || _a === void 0 ? void 0 : _a.next();
        });
    }
}
exports.ShowNextInlineSuggestionAction = ShowNextInlineSuggestionAction;
ShowNextInlineSuggestionAction.ID = commandIds_1.showNextInlineSuggestionActionId;
class ShowPreviousInlineSuggestionAction extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: ShowPreviousInlineSuggestionAction.ID,
            label: nls.localize('action.inlineSuggest.showPrevious', "Show Previous Inline Suggestion"),
            alias: 'Show Previous Inline Suggestion',
            precondition: contextkey_1.ContextKeyExpr.and(editorContextKeys_1.EditorContextKeys.writable, inlineCompletionContextKeys_1.InlineCompletionContextKeys.inlineSuggestionVisible),
            kbOpts: {
                weight: 100,
                primary: 512 /* KeyMod.Alt */ | 92 /* KeyCode.BracketLeft */,
            },
        });
    }
    run(accessor, editor) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const controller = inlineCompletionsController_1.InlineCompletionsController.get(editor);
            (_a = controller === null || controller === void 0 ? void 0 : controller.model.get()) === null || _a === void 0 ? void 0 : _a.previous();
        });
    }
}
exports.ShowPreviousInlineSuggestionAction = ShowPreviousInlineSuggestionAction;
ShowPreviousInlineSuggestionAction.ID = commandIds_1.showPreviousInlineSuggestionActionId;
class TriggerInlineSuggestionAction extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.inlineSuggest.trigger',
            label: nls.localize('action.inlineSuggest.trigger', "Trigger Inline Suggestion"),
            alias: 'Trigger Inline Suggestion',
            precondition: editorContextKeys_1.EditorContextKeys.writable
        });
    }
    run(accessor, editor) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const controller = inlineCompletionsController_1.InlineCompletionsController.get(editor);
            (_a = controller === null || controller === void 0 ? void 0 : controller.model.get()) === null || _a === void 0 ? void 0 : _a.triggerExplicitly();
        });
    }
}
exports.TriggerInlineSuggestionAction = TriggerInlineSuggestionAction;
class AcceptNextWordOfInlineCompletion extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.inlineSuggest.acceptNextWord',
            label: nls.localize('action.inlineSuggest.acceptNextWord', "Accept Next Word Of Inline Suggestion"),
            alias: 'Accept Next Word Of Inline Suggestion',
            precondition: contextkey_1.ContextKeyExpr.and(editorContextKeys_1.EditorContextKeys.writable, inlineCompletionContextKeys_1.InlineCompletionContextKeys.inlineSuggestionVisible),
            kbOpts: {
                weight: 100 /* KeybindingWeight.EditorContrib */ + 1,
                primary: 2048 /* KeyMod.CtrlCmd */ | 17 /* KeyCode.RightArrow */,
                kbExpr: contextkey_1.ContextKeyExpr.and(editorContextKeys_1.EditorContextKeys.writable, inlineCompletionContextKeys_1.InlineCompletionContextKeys.inlineSuggestionVisible),
            },
            menuOpts: [{
                    menuId: actions_1.MenuId.InlineSuggestionToolbar,
                    title: nls.localize('acceptWord', 'Accept Word'),
                    group: 'primary',
                    order: 2,
                }],
        });
    }
    run(accessor, editor) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const controller = inlineCompletionsController_1.InlineCompletionsController.get(editor);
            yield ((_a = controller === null || controller === void 0 ? void 0 : controller.model.get()) === null || _a === void 0 ? void 0 : _a.acceptNextWord(controller.editor));
        });
    }
}
exports.AcceptNextWordOfInlineCompletion = AcceptNextWordOfInlineCompletion;
class AcceptNextLineOfInlineCompletion extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.inlineSuggest.acceptNextLine',
            label: nls.localize('action.inlineSuggest.acceptNextLine', "Accept Next Line Of Inline Suggestion"),
            alias: 'Accept Next Line Of Inline Suggestion',
            precondition: contextkey_1.ContextKeyExpr.and(editorContextKeys_1.EditorContextKeys.writable, inlineCompletionContextKeys_1.InlineCompletionContextKeys.inlineSuggestionVisible),
            kbOpts: {
                weight: 100 /* KeybindingWeight.EditorContrib */ + 1,
            },
            menuOpts: [{
                    menuId: actions_1.MenuId.InlineSuggestionToolbar,
                    title: nls.localize('acceptLine', 'Accept Line'),
                    group: 'secondary',
                    order: 2,
                }],
        });
    }
    run(accessor, editor) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const controller = inlineCompletionsController_1.InlineCompletionsController.get(editor);
            yield ((_a = controller === null || controller === void 0 ? void 0 : controller.model.get()) === null || _a === void 0 ? void 0 : _a.acceptNextLine(controller.editor));
        });
    }
}
exports.AcceptNextLineOfInlineCompletion = AcceptNextLineOfInlineCompletion;
class AcceptInlineCompletion extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: commandIds_1.inlineSuggestCommitId,
            label: nls.localize('action.inlineSuggest.accept', "Accept Inline Suggestion"),
            alias: 'Accept Inline Suggestion',
            precondition: inlineCompletionContextKeys_1.InlineCompletionContextKeys.inlineSuggestionVisible,
            menuOpts: [{
                    menuId: actions_1.MenuId.InlineSuggestionToolbar,
                    title: nls.localize('accept', "Accept"),
                    group: 'primary',
                    order: 1,
                }],
            kbOpts: {
                primary: 2 /* KeyCode.Tab */,
                weight: 200,
                kbExpr: contextkey_1.ContextKeyExpr.and(inlineCompletionContextKeys_1.InlineCompletionContextKeys.inlineSuggestionVisible, editorContextKeys_1.EditorContextKeys.tabMovesFocus.toNegated(), inlineCompletionContextKeys_1.InlineCompletionContextKeys.inlineSuggestionHasIndentationLessThanTabSize, suggest_1.Context.Visible.toNegated(), editorContextKeys_1.EditorContextKeys.hoverFocused.toNegated()),
            }
        });
    }
    run(accessor, editor) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const controller = inlineCompletionsController_1.InlineCompletionsController.get(editor);
            if (controller) {
                (_a = controller.model.get()) === null || _a === void 0 ? void 0 : _a.accept(controller.editor);
                controller.editor.focus();
            }
        });
    }
}
exports.AcceptInlineCompletion = AcceptInlineCompletion;
class HideInlineCompletion extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: HideInlineCompletion.ID,
            label: nls.localize('action.inlineSuggest.hide', "Hide Inline Suggestion"),
            alias: 'Hide Inline Suggestion',
            precondition: inlineCompletionContextKeys_1.InlineCompletionContextKeys.inlineSuggestionVisible,
            kbOpts: {
                weight: 100,
                primary: 9 /* KeyCode.Escape */,
            }
        });
    }
    run(accessor, editor) {
        return __awaiter(this, void 0, void 0, function* () {
            const controller = inlineCompletionsController_1.InlineCompletionsController.get(editor);
            (0, observable_1.transaction)(tx => {
                var _a;
                (_a = controller === null || controller === void 0 ? void 0 : controller.model.get()) === null || _a === void 0 ? void 0 : _a.stop(tx);
            });
        });
    }
}
exports.HideInlineCompletion = HideInlineCompletion;
HideInlineCompletion.ID = 'editor.action.inlineSuggest.hide';
class ToggleAlwaysShowInlineSuggestionToolbar extends actions_1.Action2 {
    constructor() {
        super({
            id: ToggleAlwaysShowInlineSuggestionToolbar.ID,
            title: nls.localize('action.inlineSuggest.alwaysShowToolbar', "Always Show Toolbar"),
            f1: false,
            precondition: undefined,
            menu: [{
                    id: actions_1.MenuId.InlineSuggestionToolbar,
                    group: 'secondary',
                    order: 10,
                }],
            toggled: contextkey_1.ContextKeyExpr.equals('config.editor.inlineSuggest.showToolbar', 'always')
        });
    }
    run(accessor, editor) {
        return __awaiter(this, void 0, void 0, function* () {
            const configService = accessor.get(configuration_1.IConfigurationService);
            const currentValue = configService.getValue('editor.inlineSuggest.showToolbar');
            const newValue = currentValue === 'always' ? 'onHover' : 'always';
            configService.updateValue('editor.inlineSuggest.showToolbar', newValue);
        });
    }
}
exports.ToggleAlwaysShowInlineSuggestionToolbar = ToggleAlwaysShowInlineSuggestionToolbar;
ToggleAlwaysShowInlineSuggestionToolbar.ID = 'editor.action.inlineSuggest.toggleAlwaysShowToolbar';
