// @theia/monaco-editor-core/esm/vs/editor/contrib/inlineCompletions/browser/inlineCompletionsController.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var InlineCompletionsController_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InlineCompletionsController = void 0;
const aria_1 = require("../../../../base/browser/ui/aria/aria");
const event_1 = require("../../../../base/common/event");
const lifecycle_1 = require("../../../../base/common/lifecycle");
const observable_1 = require("../../../../base/common/observable");
const coreCommands_1 = require("../../../browser/coreCommands");
const position_1 = require("../../../common/core/position");
const languageFeatureDebounce_1 = require("../../../common/services/languageFeatureDebounce");
const languageFeatures_1 = require("../../../common/services/languageFeatures");
const commandIds_1 = require("./commandIds");
const ghostTextWidget_1 = require("./ghostTextWidget");
const inlineCompletionContextKeys_1 = require("./inlineCompletionContextKeys");
const inlineCompletionsHintsWidget_1 = require("./inlineCompletionsHintsWidget");
const inlineCompletionsModel_1 = require("./inlineCompletionsModel");
const suggestWidgetInlineCompletionProvider_1 = require("./suggestWidgetInlineCompletionProvider");
const nls_1 = require("../../../../nls");
const audioCueService_1 = require("../../../../platform/audioCues/browser/audioCueService");
const commands_1 = require("../../../../platform/commands/common/commands");
const configuration_1 = require("../../../../platform/configuration/common/configuration");
const contextkey_1 = require("../../../../platform/contextkey/common/contextkey");
const instantiation_1 = require("../../../../platform/instantiation/common/instantiation");
const keybinding_1 = require("../../../../platform/keybinding/common/keybinding");
let InlineCompletionsController = InlineCompletionsController_1 = class InlineCompletionsController extends lifecycle_1.Disposable {
    static get(editor) {
        return editor.getContribution(InlineCompletionsController_1.ID);
    }
    constructor(editor, instantiationService, contextKeyService, configurationService, commandService, debounceService, languageFeaturesService, audioCueService, _keybindingService) {
        super();
        this.editor = editor;
        this.instantiationService = instantiationService;
        this.contextKeyService = contextKeyService;
        this.configurationService = configurationService;
        this.commandService = commandService;
        this.debounceService = debounceService;
        this.languageFeaturesService = languageFeaturesService;
        this.audioCueService = audioCueService;
        this._keybindingService = _keybindingService;
        this.model = (0, observable_1.disposableObservableValue)('inlineCompletionModel', undefined);
        this.textModelVersionId = (0, observable_1.observableValue)(this, -1);
        this.cursorPosition = (0, observable_1.observableValue)(this, new position_1.Position(1, 1));
        this.suggestWidgetAdaptor = this._register(new suggestWidgetInlineCompletionProvider_1.SuggestWidgetAdaptor(this.editor, () => { var _a, _b; return (_b = (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.selectedInlineCompletion.get()) === null || _b === void 0 ? void 0 : _b.toSingleTextEdit(undefined); }, (tx) => this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other), (item) => {
            (0, observable_1.transaction)(tx => {
                var _a;
                /** @description handleSuggestAccepted */
                this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other);
                (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.handleSuggestAccepted(item);
            });
        }));
        this._enabled = (0, observable_1.observableFromEvent)(this.editor.onDidChangeConfiguration, () => this.editor.getOption(62 /* EditorOption.inlineSuggest */).enabled);
        this.ghostTextWidget = this._register(this.instantiationService.createInstance(ghostTextWidget_1.GhostTextWidget, this.editor, {
            ghostText: this.model.map((v, reader) => v === null || v === void 0 ? void 0 : v.ghostText.read(reader)),
            minReservedLineCount: (0, observable_1.constObservable)(0),
            targetTextModel: this.model.map(v => v === null || v === void 0 ? void 0 : v.textModel),
        }));
        this._debounceValue = this.debounceService.for(this.languageFeaturesService.inlineCompletionsProvider, 'InlineCompletionsDebounce', { min: 50, max: 50 });
        this._register(new inlineCompletionContextKeys_1.InlineCompletionContextKeys(this.contextKeyService, this.model));
        this._register(event_1.Event.runAndSubscribe(editor.onDidChangeModel, () => (0, observable_1.transaction)(tx => {
            /** @description onDidChangeModel */
            this.model.set(undefined, tx);
            this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other);
            const textModel = editor.getModel();
            if (textModel) {
                const model = instantiationService.createInstance(inlineCompletionsModel_1.InlineCompletionsModel, textModel, this.suggestWidgetAdaptor.selectedItem, this.cursorPosition, this.textModelVersionId, this._debounceValue, (0, observable_1.observableFromEvent)(editor.onDidChangeConfiguration, () => editor.getOption(117 /* EditorOption.suggest */).preview), (0, observable_1.observableFromEvent)(editor.onDidChangeConfiguration, () => editor.getOption(117 /* EditorOption.suggest */).previewMode), (0, observable_1.observableFromEvent)(editor.onDidChangeConfiguration, () => editor.getOption(62 /* EditorOption.inlineSuggest */).mode), this._enabled);
                this.model.set(model, tx);
            }
        })));
        const getReason = (e) => {
            var _a;
            if (e.isUndoing) {
                return inlineCompletionsModel_1.VersionIdChangeReason.Undo;
            }
            if (e.isRedoing) {
                return inlineCompletionsModel_1.VersionIdChangeReason.Redo;
            }
            if ((_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.isAcceptingPartially) {
                return inlineCompletionsModel_1.VersionIdChangeReason.AcceptWord;
            }
            return inlineCompletionsModel_1.VersionIdChangeReason.Other;
        };
        this._register(editor.onDidChangeModelContent((e) => (0, observable_1.transaction)(tx => 
        /** @description onDidChangeModelContent */
        this.updateObservables(tx, getReason(e)))));
        this._register(editor.onDidChangeCursorPosition(e => (0, observable_1.transaction)(tx => {
            var _a;
            /** @description onDidChangeCursorPosition */
            this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other);
            if (e.reason === 3 /* CursorChangeReason.Explicit */ || e.source === 'api') {
                (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.stop(tx);
            }
        })));
        this._register(editor.onDidType(() => (0, observable_1.transaction)(tx => {
            var _a;
            /** @description onDidType */
            this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other);
            if (this._enabled.get()) {
                (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.trigger(tx);
            }
        })));
        this._register(this.commandService.onDidExecuteCommand((e) => {
            // These commands don't trigger onDidType.
            const commands = new Set([
                coreCommands_1.CoreEditingCommands.Tab.id,
                coreCommands_1.CoreEditingCommands.DeleteLeft.id,
                coreCommands_1.CoreEditingCommands.DeleteRight.id,
                commandIds_1.inlineSuggestCommitId,
                'acceptSelectedSuggestion',
            ]);
            if (commands.has(e.commandId) && editor.hasTextFocus() && this._enabled.get()) {
                (0, observable_1.transaction)(tx => {
                    var _a;
                    /** @description onDidExecuteCommand */
                    (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.trigger(tx);
                });
            }
        }));
        this._register(this.editor.onDidBlurEditorWidget(() => {
            // This is a hidden setting very useful for debugging
            if (this.contextKeyService.getContextKeyValue('accessibleViewIsShown') || this.configurationService.getValue('editor.inlineSuggest.keepOnBlur') ||
                editor.getOption(62 /* EditorOption.inlineSuggest */).keepOnBlur) {
                return;
            }
            if (inlineCompletionsHintsWidget_1.InlineSuggestionHintsContentWidget.dropDownVisible) {
                return;
            }
            (0, observable_1.transaction)(tx => {
                var _a;
                /** @description onDidBlurEditorWidget */
                (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.stop(tx);
            });
        }));
        this._register((0, observable_1.autorun)(reader => {
            var _a;
            /** @description forceRenderingAbove */
            const state = (_a = this.model.read(reader)) === null || _a === void 0 ? void 0 : _a.state.read(reader);
            if (state === null || state === void 0 ? void 0 : state.suggestItem) {
                if (state.ghostText.lineCount >= 2) {
                    this.suggestWidgetAdaptor.forceRenderingAbove();
                }
            }
            else {
                this.suggestWidgetAdaptor.stopForceRenderingAbove();
            }
        }));
        this._register((0, lifecycle_1.toDisposable)(() => {
            this.suggestWidgetAdaptor.stopForceRenderingAbove();
        }));
        let lastInlineCompletionId = undefined;
        this._register((0, observable_1.autorun)(reader => {
            /** @description play audio cue & read suggestion */
            const model = this.model.read(reader);
            const state = model === null || model === void 0 ? void 0 : model.state.read(reader);
            if (!model || !state || !state.inlineCompletion) {
                lastInlineCompletionId = undefined;
                return;
            }
            if (state.inlineCompletion.semanticId !== lastInlineCompletionId) {
                lastInlineCompletionId = state.inlineCompletion.semanticId;
                const lineText = model.textModel.getLineContent(state.ghostText.lineNumber);
                this.audioCueService.playAudioCue(audioCueService_1.AudioCue.inlineSuggestion).then(() => {
                    if (this.editor.getOption(8 /* EditorOption.screenReaderAnnounceInlineSuggestion */)) {
                        this.provideScreenReaderUpdate(state.ghostText.renderForScreenReader(lineText));
                    }
                });
            }
        }));
        this._register(new inlineCompletionsHintsWidget_1.InlineCompletionsHintsWidget(this.editor, this.model, this.instantiationService));
        this._register(this.configurationService.onDidChangeConfiguration(e => {
            if (e.affectsConfiguration('accessibility.verbosity.inlineCompletions')) {
                this.editor.updateOptions({ inlineCompletionsAccessibilityVerbose: this.configurationService.getValue('accessibility.verbosity.inlineCompletions') });
            }
        }));
        this.editor.updateOptions({ inlineCompletionsAccessibilityVerbose: this.configurationService.getValue('accessibility.verbosity.inlineCompletions') });
    }
    provideScreenReaderUpdate(content) {
        const accessibleViewShowing = this.contextKeyService.getContextKeyValue('accessibleViewIsShown');
        const accessibleViewKeybinding = this._keybindingService.lookupKeybinding('editor.action.accessibleView');
        let hint;
        if (!accessibleViewShowing && accessibleViewKeybinding && this.editor.getOption(147 /* EditorOption.inlineCompletionsAccessibilityVerbose */)) {
            hint = (0, nls_1.localize)('showAccessibleViewHint', "Inspect this in the accessible view ({0})", accessibleViewKeybinding.getAriaLabel());
        }
        hint ? (0, aria_1.alert)(content + ', ' + hint) : (0, aria_1.alert)(content);
    }
    /**
     * Copies over the relevant state from the text model to observables.
     * This solves all kind of eventing issues, as we make sure we always operate on the latest state,
     * regardless of who calls into us.
     */
    updateObservables(tx, changeReason) {
        var _a, _b;
        const newModel = this.editor.getModel();
        this.textModelVersionId.set((_a = newModel === null || newModel === void 0 ? void 0 : newModel.getVersionId()) !== null && _a !== void 0 ? _a : -1, tx, changeReason);
        this.cursorPosition.set((_b = this.editor.getPosition()) !== null && _b !== void 0 ? _b : new position_1.Position(1, 1), tx);
    }
    shouldShowHoverAt(range) {
        var _a;
        const ghostText = (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.ghostText.get();
        if (ghostText) {
            return ghostText.parts.some(p => range.containsPosition(new position_1.Position(ghostText.lineNumber, p.column)));
        }
        return false;
    }
    shouldShowHoverAtViewZone(viewZoneId) {
        return this.ghostTextWidget.ownsViewZone(viewZoneId);
    }
    hide() {
        (0, observable_1.transaction)(tx => {
            var _a;
            (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.stop(tx);
        });
    }
};
exports.InlineCompletionsController = InlineCompletionsController;
InlineCompletionsController.ID = 'editor.contrib.inlineCompletionsController';
exports.InlineCompletionsController = InlineCompletionsController = InlineCompletionsController_1 = __decorate([
    __param(1, instantiation_1.IInstantiationService),
    __param(2, contextkey_1.IContextKeyService),
    __param(3, configuration_1.IConfigurationService),
    __param(4, commands_1.ICommandService),
    __param(5, languageFeatureDebounce_1.ILanguageFeatureDebounceService),
    __param(6, languageFeatures_1.ILanguageFeaturesService),
    __param(7, audioCueService_1.IAudioCueService),
    __param(8, keybinding_1.IKeybindingService)
], InlineCompletionsController);
